README.md

```diff
@@ -22,20 +43,105 @@ Or install it yourself as:
 
 ## Usage
 
-TODO: Write usage instructions here
+### Defining and Creating a Task
+
+To define a task, mix in the Task module:
+
+```ruby
+class FetchFile
+  include Task::Task
+
+  data_attr_reader :external_host
+  data_attr_reader :filename
+end
+```
+
+A task contains a set of data key-value pairs; The `data_attr_reader` helper provides accessors for various expected
+data fields, but is not required.
+
+To build a task, we invoke the build method:
+
+```ruby
+file_task = FetchFile.build(
+  id: 'file1',
+  task_list: 'datto.com',
+  external_host: 'datto.com',
+  filename: 'file1.txt')
+```
+
+Here, the task belongs to the 'datto.com' task list. The ID of the task should be unique across the task list (if one
+is not specified, a UUID will be used). Other provided fields just become part of the data of task; this is just syntactic
+sugar for:
+
+```ruby
+file_task = FetchFile.new(
+  id: 'file1',
+  task_list: 'datto.com',
+  data: { external_host: 'datto.com', filename: 'file1.txt' })
+```
 
-## Development
+On the task object, the `data_attr_readers` allow access to data fields, so `file_task.data[:external_host]` and
+`file_task.external_host` are equivalent.
 
-After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake test` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.
+### Saving and Loading Tasks from the Datastore
 
-To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).
+To save to the datastore:
+
+```ruby
+file_task.save
+```
+
+To fetch a particular task (by id) from the datastore:
+
+```ruby
+file_task = Task::Task.find(task_list, id)
+```
+
+To fetch all tasks for a task list from the datastore:
+
+```ruby
+tasks_enumerator = Task::Task.all(task_list)
+```
+
+### Serializing and Deserializing Tasks
+
+To serialize as a hash:
+
+```ruby
+task_hash = my_task.as_hash
+```
+
+To deserialize a task hash to the task that was originally serialized:
+
+```ruby
+my_task = Task::Task.from_hash(task_hash)
+```
+
+Task does not enforce a particular over the wire serialization format.
+
+### Configuring the Datastore
+
+The datastore interface is backed by an adapter. Which adapter is used, and how it is
+constructed, is configurable, by specifying the adapter builder.
+
+```ruby
+Task::DataInterface::Interface.adapter_builder = ->(_options) do
+  session = Cassandra.cluster(port: 1234, hosts: ['my_host']).connect('my_tasks')
+  CassandraAdapter.new(client: Cassava::Client.new(session))
```

Comment:
is it worth mentioning here that we expect the last line of this delegate to return the adapter to the datastore?